docs/docs/best-practices.md

```diff
@@ -0,0 +1,168 @@
+
+ ![](./imgs/simple_trace.png)
+
+ As you can see, the output gets rendered as a list in the `output` column. We can unpack the list to render the output on individual columns. To do so, we need to return a dict with the name of each attribute.
+
+
+```python
+@weave.op
+def compute_some_values(a: int, b: int, coolness: str):
+    "Simple function to compute some values"
+    total = a + b
+    is_divisible = a % b == 0
+    return {"coolness": coolness, "total": total, "is_divisible": is_divisible}
+```
+![](./imgs/dict_output_trace.png)
+
+It is up to you if you want to change the output or not, but it may be useful to have the output in a dict. Also, most of the LLMs APIs return some kind of structured output, so it is a good idea to return a dict on other parts of your codebase too.
+
+## Weave uses Pydantic
+
+bla bla bla about Pydantic and how Weave uses Pydantic BaseModel everywhere
+
+### weave.Object and weave.Model
+
+A `weave.Model` is a `BaseModel` with ... some explanation about how to use it. We should be explicit about the `weave.Model` and attributes like `generate`, `predict`, `__call__`, etc...
+
+### model_validator and delayed init
+
+We need to document the `model_validator` decorator and when to use it. Typically, we use it to validate the data in the `__init__` method of a Pydantic BaseModel. For LLM use cases this is a good place to define the client or load the model weights.
+
+Let's consider the following example:
+
+```python
+import json
+import openai
+import weave
+
+class LLMModel(weave.Model):
+    model_name: str
+    prompt_template: str
+
+    @weave.op()
+    async def predict(self, sentence: str) -> dict:
+        client = openai.AsyncClient()
+
+        response = await client.chat.completions.create(
+            model=self.model_name,
+            messages=[
+                {"role": "user", "content": self.prompt_template.format(sentence=sentence)}
+            ],
+        )
+        return response.choices[0].message.content
+```
+Everytime we call `predict`, we create a new `openai.AsyncClient`. This is not efficient and can lead to a problems when we have a lot of requests to the model and the way Python handles processes. An alternative to this, is using the `model_validator` decorator to create the client before the model is initialized. As a `weave.Model` is a subclass of `pydantic.BaseModel`, we can use the `model_validator` decorator to create the client before the model is initialized.
+
+```python
+import json
+import openai
+import weave
+# highlight-next-line
+from pydantic import Field, model_validator
+
+class LLMModel(weave.Model):
+    model_name: str
+    prompt_template: str
+    client: openai.AsyncClient = Field(init=False)
+    
+    # highlight-next-line
+    @model_validator(mode="before")
+    def create_client(cls, values: dict):
+        values['client'] = openai.AsyncClient()
+        return values
```

Comment:
I think in general we should discourage using something like `@model_validator` since it's more likely to break things.

This might be more simply expressed as:
```
_client: openai.AsyncClient = PrivateAttr(default_factory=openai.AsyncClient)
```

Reply:
@andrewtruong how would you give people the possibility of having a constructor? If people don't want to pass in the finished objects to the model the question is what pydantic option to use to allow users some pydantic-compatible constructor to create their specific model attributes (either just a client, or an actual loaded in model, RAG model, etc.)?

Reply:
I'm not sure I understand what you mean by this.  Can you give me an example?

Reply:
The code in the below thread for example - instead of passing in the ready model object you pass in the model name and init the model object dynamically as an attribute (either through any openai, anthropic, litellm client or as an actual local model you load in).

Reply:
Yep see the example below